web/src/repo/TreePage.tsx

```diff
@@ -149,272 +150,234 @@ interface Props
         } else {
             eventLogger.logViewEvent('Tree')
         }
-    }
+    }, [filePath])
+
+    const [showOlderCommits, setShowOlderCommits] = useState(false)
 
-    public componentDidMount(): void {
-        this.subscriptions.add(
-            this.componentUpdates
-                .pipe(
-                    distinctUntilChanged(
-                        (x, y) =>
-                            x.repoName === y.repoName &&
-                            x.rev === y.rev &&
-                            x.commitID === y.commitID &&
-                            x.filePath === y.filePath
-                    ),
-                    tap(props => this.logViewEvent(props)),
-                    switchMap(props =>
-                        fetchTreeEntries({
-                            repoName: props.repoName,
-                            commitID: props.commitID,
-                            rev: props.rev,
-                            filePath: props.filePath,
-                            first: 2500,
-                        }).pipe(
-                            catchError(err => [asError(err)]),
-                            map(c => ({ treeOrError: c })),
-                            startWith<Pick<State, 'treeOrError'>>({ treeOrError: undefined })
-                        )
-                    )
-                )
-                .subscribe(
-                    stateUpdate => this.setState(stateUpdate),
-                    err => console.error(err)
-                )
+    const onShowOlderCommitsClicked = useCallback(
+        (e: React.MouseEvent): void => {
+            e.preventDefault()
+            setShowOlderCommits(true)
+        },
+        [setShowOlderCommits]
+    )
+
+    const treeOrError = useObservable(
+        useMemo(
+            () =>
+                fetchTreeEntries({
+                    repoName,
+                    commitID,
+                    rev,
+                    filePath,
+                    first: 2500,
+                }).pipe(catchError((err): [ErrorLike] => [asError(err)])),
+            [repoName, commitID, rev, filePath]
         )
-        this.componentUpdates.next(this.props)
-    }
+    )
 
-    public componentDidUpdate(): void {
-        this.componentUpdates.next(this.props)
+    const getPageTitle = (): string => {
+        const repoStr = displayRepoName(repoName)
+        if (filePath) {
+            return `${basename(filePath)} - ${repoStr}`
+        }
+        return `${repoStr}`
     }
 
-    public componentWillUnmount(): void {
-        this.subscriptions.unsubscribe()
+    const queryCommits = (args: { first?: number }): Observable<GQL.IGitCommitConnection> => {
+        const after: string | undefined = showOlderCommits ? undefined : formatISO(subYears(Date.now(), 1))
+        return fetchTreeCommits({
+            ...args,
+            repo: repoID,
+            revspec: rev || '',
+            filePath,
+            after,
+        })
     }
```

Comment:
This should be wrapped in `useCallback()` because it is passed in a JSX prop
```suggestion
    const queryCommits = useCallback((args: { first?: number }): Observable<GQL.IGitCommitConnection> => {
        const after: string | undefined = showOlderCommits ? undefined : formatISO(subYears(Date.now(), 1))
        return fetchTreeCommits({
            ...args,
            repo: repoID,
            revspec: rev || '',
            filePath,
            after,
        })
    })
```